// amocrm-js/dist/plugins/hasUpdatableTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasUpdatableTags = exports.getTaggedConstructors = void 0;
const util_1 = require("../util");
const hasUpdatableTags_1 = require("./factories/hasUpdatableTags");
function getTaggedConstructors(constructors) {
    const mixins = [hasUpdatableTags_1.hasUpdatableTags];
    const factories = Object.assign(Object.assign({}, constructors.factories), { leads: (0, util_1.applyMixins)(constructors.factories.leads, mixins), contacts: (0, util_1.applyMixins)(constructors.factories.contacts, mixins) });
    return Object.assign(Object.assign({}, constructors), { factories });
}
exports.getTaggedConstructors = getTaggedConstructors;
function hasUpdatableTags(Base) {
    return class ClientWithUpdatableTags extends Base {
        constructor(options) {
            super(options);
            const { factories } = this.constructors;
            this.leads = new factories.leads(this);
            this.contacts = new factories.contacts(this);
            this.Lead = this.assignEntity(this.leads);
            this.Contact = this.assignEntity(this.contacts);
        }
        get constructors() {
            return getTaggedConstructors(super.constructors);
        }
    };
}
exports.hasUpdatableTags = hasUpdatableTags;
