// amocrm-js/dist/api/activeRecords/Customer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Customer = exports.BaseCustomer = void 0;
const tslib_1 = require("tslib");
const ResourceEntity_1 = tslib_1.__importDefault(require("../ResourceEntity"));
const util_1 = require("../../util");
const hasCreate_1 = require("./mixins/hasCreate");
const hasUpdate_1 = require("./mixins/hasUpdate");
const hasSave_1 = require("./mixins/hasSave");
const hasFetch_1 = require("./mixins/hasFetch");
class BaseCustomer extends ResourceEntity_1.default {
    getAttributes() {
        return {
            name: this.name,
            next_price: this.next_price,
            next_date: this.next_date,
            responsible_user_id: this.responsible_user_id,
            periodicity: this.periodicity,
            created_by: this.created_by,
            updated_by: this.updated_by,
            created_at: this.created_at,
            updated_at: this.updated_at,
            custom_fields_values: this.custom_fields_values,
            _embedded: this._embedded
        };
    }
    setAttributes(attributes = {}) {
        this.id = attributes.id;
        this.name = attributes.name;
        this.next_price = attributes.next_price;
        this.responsible_user_id = attributes.responsible_user_id;
        this.periodicity = attributes.periodicity;
        this.created_by = attributes.created_by;
        this.updated_at = attributes.updated_at;
        this.created_at = attributes.created_at;
        this.custom_fields_values = attributes.custom_fields_values;
        this._embedded = attributes._embedded;
    }
}
exports.BaseCustomer = BaseCustomer;
exports.Customer = (0, util_1.applyMixins)(BaseCustomer, [
    hasCreate_1.hasCreate,
    hasUpdate_1.hasUpdate,
    hasSave_1.hasSave,
    hasFetch_1.hasFetch
]);
