// amocrm-js/dist/Client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const tslib_1 = require("tslib");
const EventEmitter_1 = require("./common/EventEmitter");
const Connection_1 = require("./common/Connection");
const Environment_1 = require("./common/Environment");
const ClientRequest_1 = require("./common/ClientRequest");
const Auth_1 = require("./common/Auth");
const Token_1 = require("./common/Token");
const constructors_1 = tslib_1.__importDefault(require("./common/constructors"));
/**
 * Основной класс библиотеки
 * */
class Client extends EventEmitter_1.EventEmitter {
    constructor(options) {
        super();
        if (!options) {
            throw new Error('NO_OPTIONS');
        }
        this.environment = new Environment_1.Environment(options);
        this.token = new Token_1.Token(this.environment);
        this.auth = new Auth_1.Auth(this.environment, this.token);
        this.connection = new Connection_1.Connection(this.environment, this.token, this.auth);
        this.request = new ClientRequest_1.ClientRequest(this.connection);
        const { factories } = this.constructors;
        this.leads = new factories.leads(this);
        this.contacts = new factories.contacts(this);
        this.companies = new factories.companies(this);
        this.Lead = this.assignEntity(this.leads);
        this.Contact = this.assignEntity(this.contacts);
        this.Company = this.assignEntity(this.companies);
    }
    get constructors() {
        return constructors_1.default;
    }
    getRequest() {
        return this.request;
    }
    getFactoryConstructors() {
        return this.constructors.factories;
    }
    getEntityConstructors() {
        return this.constructors.entities;
    }
    /**
     * Привязывает конструктор сущностей
     * @param factory - фабрика сущностей
     * @returns функция конструктор для вызова new client[Entity]
     * */
    assignEntity(factory) {
        return function (attributes) {
            return factory.from(attributes);
        };
    }
}
exports.Client = Client;
