// amocrm-js/dist/api/factories/mixins/hasTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasTags = void 0;
function hasTags(Base) {
    return class HasTags extends Base {
        get tagList() {
            if (this._tags) {
                return this._tags;
            }
            this._tags = {
                get: this.getTags.bind(this),
                create: this.createTags.bind(this),
            };
            return this._tags;
        }
        getTagsFactory() {
            if (this._tagsFactory !== undefined) {
                return this._tagsFactory;
            }
            const entityType = this.getEmbeddedKey();
            const client = this.getClient();
            const constructors = client.getFactoryConstructors();
            const factory = new constructors.tags(client, entityType);
            this._tagsFactory = factory;
            return factory;
        }
        getTags(criteria, options) {
            const factory = this.getTagsFactory();
            return factory.get(criteria, options);
        }
        createTags(criteria, options) {
            const factory = this.getTagsFactory();
            return factory.create(criteria, options);
        }
    };
}
exports.hasTags = hasTags;
