// amocrm-js/dist/api/factories/CustomerFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomerFactory = exports.BaseCustomerFactory = void 0;
const tslib_1 = require("tslib");
const util_1 = require("../../util");
const hasGetByCriteria_1 = require("./mixins/hasGetByCriteria");
const hasGetById_1 = require("./mixins/hasGetById");
const hasCreate_1 = require("./mixins/hasCreate");
const hasUpdate_1 = require("./mixins/hasUpdate");
const Customer_1 = require("../activeRecords/Customer");
const v4_1 = tslib_1.__importDefault(require("../../schema/v4"));
const ResourceFactory_1 = tslib_1.__importDefault(require("../ResourceFactory"));
class BaseCustomerFactory extends ResourceFactory_1.default {
    getEntityClass() {
        return Customer_1.Customer;
    }
    getBaseUrl() {
        return v4_1.default.entities.customers.path;
    }
    getEmbeddedKey() {
        return 'customers';
    }
    setMode() {
    }
}
exports.BaseCustomerFactory = BaseCustomerFactory;
exports.CustomerFactory = (0, util_1.applyMixins)(BaseCustomerFactory, [
    hasGetByCriteria_1.hasGetByCriteria,
    hasGetById_1.hasGetById,
    hasCreate_1.hasCreate,
    hasUpdate_1.hasUpdate
]);
