// amocrm-js/dist/api/factories/TagFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagFactory = exports.BaseTagFactory = void 0;
const tslib_1 = require("tslib");
/**
 * Фабрика для создания сделок {@link Tag}
 * */
const ResourceFactory_1 = tslib_1.__importDefault(require("../ResourceFactory"));
const v4_1 = tslib_1.__importDefault(require("../../schema/v4"));
const hasGetByCriteria_1 = require("./mixins/hasGetByCriteria");
const hasCreate_1 = require("./mixins/hasCreate");
const util_1 = require("../../util");
const Tag_1 = require("../activeRecords/Tag");
/**
 * Фабрика управления сделками
 * */
class BaseTagFactory extends ResourceFactory_1.default {
    constructor(client, entityType) {
        super(client);
        this.entityType = entityType;
    }
    getEntityClass() {
        return Tag_1.Tag;
    }
    getBaseUrl() {
        return v4_1.default.apiUrl + '/' + this.entityType + '/tags';
    }
    getEmbeddedKey() {
        return 'tags';
    }
}
exports.BaseTagFactory = BaseTagFactory;
exports.TagFactory = (0, util_1.applyMixins)(BaseTagFactory, [
    hasGetByCriteria_1.hasGetByCriteria,
    hasCreate_1.hasCreate,
]);
