// amocrm-js/dist/api/activeRecords/common/CriteriaBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CriteriaBuilder = void 0;
class CriteriaBuilder {
    constructor(entity) {
        this.entity = entity;
        this.items = [];
    }
    add(item) {
        this.items.push(item);
    }
    getCreateCriteria() {
        const attributes = this.entity.getAttributes();
        return this.items.reduce((attributes, item) => {
            return this.merge(attributes, item.getCreateCriteria());
        }, attributes);
    }
    getUpdateCriteria() {
        const attributes = this.entity.getAttributes();
        return this.items.reduce((attributes, item) => {
            return this.merge(attributes, item.getUpdateCriteria());
        }, attributes);
    }
    merge(source, criteria) {
        const { _embedded } = criteria;
        return Object.assign(Object.assign(Object.assign({}, source), criteria), { _embedded: Object.assign(Object.assign({}, source._embedded), _embedded) });
    }
}
exports.CriteriaBuilder = CriteriaBuilder;
